### PR TITLE
Adding event and error selector fields

### LINKED ISSUE #12581: Extend the language to have an EVENT.selector and ERROR.selector
## Abstract

Add `ev.selector` and `err.selector` fields that would work similarly to `fun.selector`, extracting event signature (topic 0) from its definition

```
interface MyInterface {
    event someEvent();
    error someError();
    function someFunction() external;
}

contract Test {
    function nameIsIrrelevant() public {
        bytes4  x = MyInterface.someFunction.selector; // This works
        bytes32 y = MyInterface.someEvent.selector; // This doesn't, that is what I'd like to add
        bytes32 z = MyInterface.someError.selector; // This doesn't, that is what I'd like to add
    }
}
```

## Motivation

Some cross chain mechanism, such as polygon's [FxBaseRootTunnel](https://github.com/fx-portal/contracts/blob/main/contracts/tunnel/FxBaseRootTunnel.sol), perform event inclusion check that require the event signature. This is likely to become a more frequent need with inclusion proofs being an important part of rollups.

## Specification

Add a bytes32 field named `.selector`, `.topic0` or `.signature` to events (and a similar one to errors), that the compiler would replace with the event's signature (keccak256 of the event description)

## Backwards Compatibility

I can't imagine this would cause any backward compatibility issue

### PR DESCRIPTION
**Pull request by nishant-sachdeva**
closes #12581

### CODE REVIEW

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `480595b4fd37c04c6610f961154c783bbe9994fd`:
```
Error: Trailing whitespace found:
test/libsolidity/semanticTests/error/error_selector.sol:23:    

```
Please check that your changes are working as intended.

File: libsolidity/ast/AST.h [hunk 1]
```diff
@@ -1216,7 +1216,7 @@ class EventDefinition: public CallableDeclaration, public StructurallyDocumented
 	FunctionTypePointer functionType(bool /*_internal*/) const override;
 
 	bool isVisibleInDerivedContracts() const override { return true; }
-	bool isVisibleViaContractTypeAccess() const override { return false; /* TODO */ }
+	bool isVisibleViaContractTypeAccess() const override { return true; /* TODO */ }
```
Comment (hrkrshnn) [inline libsolidity/ast/AST.h]: Can you add a test like this:
```solidity
contract C {
    event E();
}

contract Test {
   function f() ... {
         emit C.E();
   }
}
```

File: libsolidity/ast/Types.cpp [hunk 2]
```diff
@@ -3335,6 +3335,8 @@ MemberList::MemberMap FunctionType::nativeMembers(ASTNode const* _scope) const
 	}
 	case Kind::Error:
 		return {{"selector", TypeProvider::fixedBytes(4)}};
+	case Kind::Event:
+		return {{"selector", TypeProvider::fixedBytes(4)}};
```
Comment (hrkrshnn) [inline libsolidity/ast/Types.cpp]: The event selector is actually 32 bytes!

File: test/libsolidity/semanticTests/events/event_selector.sol [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+contract C is D {
+    function test1() external pure returns (bytes4, bytes4, bytes4, bytes4) {
+        assert(L.E.selector == T.E.selector);
+        assert(I.E.selector == L.E.selector);
+        assert(Y.E.selector == L.E.selector);
+
+        assert(L.E.selector != S.E.selector);
+        assert(T.E.selector != S.E.selector);
+        assert(I.E.selector != S.E.selector);
+        assert(Y.E.selector != S.E.selector);
```
Comment (hrkrshnn) [inline test/libsolidity/semanticTests/events/event_selector.sol]: For anonymous events, `Y.E`, selector does not make sense. So `Y.E.selector` should be a type error.

File: libsolidity/ast/Types.cpp [hunk 4]
```diff
@@ -3335,6 +3335,12 @@ MemberList::MemberMap FunctionType::nativeMembers(ASTNode const* _scope) const
 	}
 	case Kind::Error:
 		return {{"selector", TypeProvider::fixedBytes(4)}};
+	case Kind::Event:
+	{
+		if(!((dynamic_cast<EventDefinition const&>(declaration())).isAnonymous()))
```
Comment (stackenbotten) [inline libsolidity/ast/Types.cpp]: Coding style error

File: test/libsolidity/semanticTests/events/event_selector.sol [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    function test3() external returns (bytes32) {
+        return (F.selector);
+    }
+}
+// ====
+// compileViaYul: also
+// ----
+// test1() -> 0x92bbf6e800000000000000000000000000000000000000000000000000000000, 0x2ff06700000000000000000000000000000000000000000000000000000000, 0x92bbf6e800000000000000000000000000000000000000000000000000000000
```
Comment (hrkrshnn) [inline test/libsolidity/semanticTests/events/event_selector.sol]: This is incorrect. It should be 32 bytes.

File: test/libsolidity/semanticTests/error/selector.sol [hunk 6]
```diff
@@ -1,20 +0,0 @@
-library L {
```
Comment (chriseth) [inline test/libsolidity/semanticTests/error/selector.sol]: Can you please just keep the test? Some duplication does not hurt.

File: Changelog.md [hunk 7]
```diff
@@ -1,6 +1,7 @@
 ### 0.8.14 (unreleased)
 
 Language Features:
+* General: Extended the language to add the event selector field.
```
Comment (chriseth) [inline Changelog.md]: All the items in this section extend the language, so it does not have to be explicitly mentioned.

File: libsolidity/ast/AST.h (hunk 1 above)
Comment (chriseth) [inline libsolidity/ast/AST.h]: Please remove the todo when you are confident that your change is correct and is not a breaking change.

Comment (chriseth) [reply libsolidity/ast/AST.h]: Also add this to the changelog if it is a change.

File: libsolidity/ast/Types.cpp [hunk 8]
```diff
@@ -3335,6 +3335,12 @@ MemberList::MemberMap FunctionType::nativeMembers(ASTNode const* _scope) const
 	}
 	case Kind::Error:
 		return {{"selector", TypeProvider::fixedBytes(4)}};
+	case Kind::Event:
+	{
+		if (!((dynamic_cast<EventDefinition const&>(declaration())).isAnonymous()))
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Can you reduce the number of parentheses?

File: libsolidity/ast/Types.h [hunk 9]
```diff
@@ -1496,6 +1501,7 @@ class FunctionType: public Type
 	StateMutability m_stateMutability = StateMutability::NonPayable;
 	Declaration const* m_declaration = nullptr;
 	Options const m_options;
+
```
Comment (chriseth) [inline libsolidity/ast/Types.h]: Please avoid adding unrelated changes.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 10]
```diff
@@ -1774,6 +1774,10 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
 			else if (
 				functionType.kind() == FunctionType::Kind::Declaration ||
 				functionType.kind() == FunctionType::Kind::Error ||
+				(
+					functionType.kind() == FunctionType::Kind::Event &&
+					!((dynamic_cast<EventDefinition const&>(functionType.declaration())).isAnonymous())
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Please reduce the number of parantheses.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 11]
```diff
@@ -1782,6 +1786,7 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
 				solAssert(functionType.hasDeclaration());
 				solAssert(
 					functionType.kind() == FunctionType::Kind::Error ||
+					functionType.kind() == FunctionType::Kind::Event ||
```
Comment (hrkrshnn) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: I think this needs to be changed. This branch treats `error` and `event` similarly. events have to be treated differently because their signature is 32 bytes.

Comment (hrkrshnn) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Also to add, there should be an assert that the event is not anonymous.

File: libsolidity/ast/AST.h (hunk 1 above)
Comment (nishant-sachdeva) [reply libsolidity/ast/AST.h]: The todo was there before I took over the issue. So, I left it as is. 

But I'll remove it and add an entry to the changelog for the same

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `a2f48bb05628f42f4a7b84d87393c005c499ccd8`:
```
Error: Trailing whitespace found:
Changelog.md:4:* General: Add the event selector field. 

```
Please check that your changes are working as intended.

File: Changelog.md [hunk 12]
```diff
@@ -1,6 +1,8 @@
 ### 0.8.14 (unreleased)
 
 Language Features:
+* General: Add the event selector field.
+* General: Errors and Events are visible via other contracts.
```
Comment (chriseth) [inline Changelog.md]:
```suggestion
* General: Errors and Events allow qualified access from other contracts.
```
(we should explain what that means in the release post)

File: Changelog.md [hunk 13]
```diff
@@ -1,6 +1,8 @@
 ### 0.8.14 (unreleased)
 
 Language Features:
+* General: Add the event selector field.
```
Comment (chriseth) [inline Changelog.md]:
```suggestion
* General: Add `E.selector` for a non-anonymous event `E` to access the 32-byte selector topic.
```

**Comment by chriseth** [comment]
This needs a documentation change: Members of events.

File: test/libsolidity/semanticTests/error/error_selector.sol [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    error E();
+    function f() external pure;
+}
+
+contract D {
+    error F();
+}
+
+contract C is D {
+    function test1() public pure returns (bytes4, bytes4, bytes4, bytes4) {
```
Comment (wechman) [inline test/libsolidity/semanticTests/error/error_selector.sol:23]: Shouldn't the error selector be 32 bytes long?

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 15]
```diff
@@ -1777,8 +1782,10 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
 			if (functionType.kind() == FunctionType::Kind::External)
 				CompilerUtils(m_context).popStackSlots(functionType.sizeOnStack() - 2);
 			m_context << Instruction::SWAP1 << Instruction::POP;
-			/// need to store it as bytes4
-			utils().leftShiftNumberOnStack(224);
+			if (!(functionType.kind() == FunctionType::Kind::Event))
+				/// need to store it as bytes4.
+				/// An event selector is 32 bytes, so this left shift is not required in that case
+				utils().leftShiftNumberOnStack(224);
```
Comment (wechman) [inline libsolidity/codegen/ExpressionCompiler.cpp]: This line is not up to date with above comment.

File: docs/units-and-global-variables.rst [hunk 16]
```diff
@@ -161,6 +161,13 @@ Members of string
 
 - ``string.concat(...) returns (string memory)``: :ref:`Concatenates variable number of string arguments to one string array<string-concat>`
 
+.. index:: event members
+
+Members of event
+----------------
+
+- ``event.selector()``: The signature is defined as the canonical expression of the basic prototype without data location specifier. It is the first (left, high-order in big-endian) 32 bytes of the Keccak-256 hash of the signature of the event.
```
Comment (wechman) [inline docs/units-and-global-variables.rst]: `first 32 bytes of the Keccak-256` means, the whole Keccak-256 hash, right?

Comment (nishant-sachdeva) [reply docs/units-and-global-variables.rst]: Yes. Changed it in the latest push.

File: docs/units-and-global-variables.rst [hunk 17]
```diff
@@ -161,6 +161,13 @@ Members of string
 
 - ``string.concat(...) returns (string memory)``: :ref:`Concatenates variable number of string arguments to one string array<string-concat>`
 
+.. index:: event members
```
Comment (chriseth) [inline docs/units-and-global-variables.rst]: This section is about global variables. You need to add this documentation to the type documentation for events, or just to the documentation for events in general.

File: test/libsolidity/semanticTests/events/simple.sol [hunk 18]
```diff
@@ -0,0 +1,14 @@
+contract C {
+    event E();
+}
+
+contract Test {
+    function f() public {
+        emit C.E();
```
Comment (chriseth) [inline test/libsolidity/semanticTests/events/simple.sol:8]: Could you make this a bit more complicated by adding another event called `E` to `Test`, but which has a different signature? Just to check that the correct event is emitted and that the test infrastructure can also cope with that.

File: test/libsolidity/syntaxTests/events/event_selector_syntax.sol [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+    event E(uint);
+}
+library T {
+    event E();
+}
+interface I {
+    event E();
+}
+library Y {
+    event E() anonymous;
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/events/event_selector_syntax.sol]: This needs to be its own test as well.

**Review by chriseth — CHANGES_REQUESTED** [review]

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 20]
```diff
@@ -1774,10 +1779,11 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
 		if (member == "selector")
 		{
 			auto const& functionType = dynamic_cast<FunctionType const&>(*_memberAccess.expression().annotation().type);
+			solAssert(!(functionType.kind() == FunctionType::Kind::Event));
+
 			if (functionType.kind() == FunctionType::Kind::External)
 				CompilerUtils(m_context).popStackSlots(functionType.sizeOnStack() - 2);
 			m_context << Instruction::SWAP1 << Instruction::POP;
-			/// need to store it as bytes4
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp:1781]: Please keep this.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 21]
```diff
@@ -1774,10 +1779,11 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
 		if (member == "selector")
 		{
 			auto const& functionType = dynamic_cast<FunctionType const&>(*_memberAccess.expression().annotation().type);
+			solAssert(!(functionType.kind() == FunctionType::Kind::Event));
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp:1784]: Please add a comment: All events should already be caught by the shortcut above.

File: libsolidity/ast/AST.h [hunk 22]
```diff
@@ -1216,7 +1216,7 @@ class EventDefinition: public CallableDeclaration, public StructurallyDocumented
 	FunctionTypePointer functionType(bool /*_internal*/) const override;
 
 	bool isVisibleInDerivedContracts() const override { return true; }
-	bool isVisibleViaContractTypeAccess() const override { return false; /* TODO */ }
+	bool isVisibleViaContractTypeAccess() const override { return true; }
```
Comment (hrkrshnn) [inline libsolidity/ast/AST.h:1219]: I think now we really will have to export used events in the ABI: https://github.com/ethereum/solidity/pull/10996

**Comment by chriseth** [comment]
What do you think about the remaining comments from 7 days ago, @nishant-sachdeva ?

File: test/libsolidity/syntaxTests/events/event_selector_library.sol [hunk 23]
```diff
@@ -0,0 +1,17 @@
...
+    }
+
+    bytes32 s5 = Y.E.selector;
+
+    function test2() view external returns (bytes32) {
+        return s5;
+    }
+}
+// ----
+// TypeError 9582: (123-135): Member "selector" not found or not visible after argument-dependent lookup in function ().
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/events/event_selector_library.sol]: Is it on purpose that `.selector` is used twice but there is only a single error? Which one are you testing? Is one of them working?

File: test/libsolidity/syntaxTests/events/event_selector_library.sol [hunk 24]
```diff
@@ -0,0 +1,17 @@
+library Y {
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/events/event_selector_library.sol]: I think the name of the test should include the word 'anonymous' since this is what is being tested.

Please also add more tests for different ways to access the selector of an anonymous event.

File: test/libsolidity/syntaxTests/events/event_selector_library.sol (hunk 23 above)
Comment (nishant-sachdeva) [reply test/libsolidity/syntaxTests/events/event_selector_library.sol]: Both the selector calls should not be working. I've confirmed this after removing each function one by one and checking for syntax errors in the other.

But it's printing only 1, presumably because the code stops after finding the first syntax error.

I'll split the tests into two files and push changes

File: docs/contracts/events.rst [hunk 25]
```diff
@@ -136,6 +136,13 @@ The output of the above looks like the following (trimmed):
        }
     }
 
+
+Members of event
```
Comment (ekpyron) [inline docs/contracts/events.rst]:
```suggestion
Members of Events
```
maybe, for consistency with the rest of the file?

File: docs/contracts/events.rst [hunk 26]
```diff
@@ -136,6 +136,13 @@ The output of the above looks like the following (trimmed):
...
+================
+
+- ``event.selector()``: The signature is defined as the canonical expression of the basic prototype without data location specifier. It is the the Keccak-256 hash of the signature of the event.
```
Comment (ekpyron) [inline docs/contracts/events.rst]: What's the reason for the parentheses here?

Comment (nishant-sachdeva) [reply docs/contracts/events.rst]: That's a mistake. Removed in the latest push.

File: docs/contracts/events.rst [hunk 27]
```diff
@@ -136,6 +136,13 @@ The output of the above looks like the following (trimmed):
        }
     }
 
+
+Members of events
```
Comment (ekpyron) [inline docs/contracts/events.rst]:
```suggestion
Members of Events
```

File: docs/units-and-global-variables.rst [hunk 28]
```diff
@@ -161,6 +161,7 @@ Members of string
...
```
Comment (ekpyron) [inline docs/units-and-global-variables.rst]: Should this be here?

Comment (nishant-sachdeva) [reply docs/units-and-global-variables.rst]: Nope. Will remove it in the latest push

File: test/libsolidity/syntaxTests/events/event_selector_library_declared_outside_but_called_inside_function.sol [hunk 29]
```diff
@@ -0,0 +1,13 @@
...
+
+contract C {
+    bytes32 s5 = Y.E.selector;
+
+    function test2() view external returns (bytes32) {
+        return s5;
+    }
+}
+// ----
+// TypeError 9582: (70-82): Member "selector" not found or not visible after argument-dependent lookup in function ().
```
Comment (hrkrshnn) [inline test/libsolidity/syntaxTests/events/event_selector_library_declared_outside_but_called_inside_function.sol:13]: > function ().

Nitpick: need to fix the error message. I think there is a PR already for this.

Comment (hrkrshnn) [reply test/libsolidity/syntaxTests/events/event_selector_library_declared_outside_but_called_inside_function.sol:13]: https://github.com/ethereum/solidity/pull/13005 is the PR

File: docs/units-and-global-variables.rst [hunk 30]
```diff
@@ -161,7 +161,6 @@ Members of string
 
 - ``string.concat(...) returns (string memory)``: :ref:`Concatenates variable number of string arguments to one string array<string-concat>`
 
-
```
Comment (ekpyron) [inline docs/units-and-global-variables.rst]: While I guess we could remove the newline here, it'd be nicer to keep the PR clean of unrelated changes like this :-).

**Review by ekpyron — DISMISSED** [review]
Looks good to me now. If we merge we should remember rebasing https://github.com/ethereum/solidity/pull/13005 afterwards, though, since it may change the error messages of the new tests.

File: docs/contracts/events.rst [hunk 31]
```diff
@@ -136,6 +136,13 @@ The output of the above looks like the following (trimmed):
        }
     }
 
+
+Members of Events
+=================
+
+- ``event.selector``: The signature is defined as the canonical expression of the basic prototype without data location specifier. It is the the Keccak-256 hash of the signature of the event.
```
Comment (chriseth) [inline docs/contracts/events.rst]: I'll change this before merging.

**Review by chriseth — APPROVED** [review]
